dig_test.go

```diff
@@ -0,0 +1,360 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dig
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestEndToEndSuccess(t *testing.T) {
+	t.Parallel()
+
+	t.Run("pointer", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(&bytes.Buffer{}), "provide failed")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			require.NotNil(t, b, "invoke got nil buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("nil pointer", func(t *testing.T) {
+		// Dig shouldn't forbid this - it's perfectly reasonable to explicitly
+		// provide a typed nil, since that's often a convenient way to supply a
+		// default no-op implementation.
+		c := New()
+		require.NoError(t, c.Provide((*bytes.Buffer)(nil)), "provide failed")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			require.Nil(t, b, "expected to get nil buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("pointer constructor", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() *bytes.Buffer {
+			return &bytes.Buffer{}
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			require.NotNil(t, b, "invoke got nil buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("struct", func(t *testing.T) {
+		t.Skip("Not yet supported.")
+		c := New()
+		buf := &bytes.Buffer{}
+		buf.WriteString("foo")
+		require.NoError(t, c.Provide(*buf), "provide failed")
+		require.NoError(t, c.Invoke(func(b bytes.Buffer) {
+			// ensure we're getting back the buffer we put in
+			require.Equal(t, "foo", buf.String(), "invoke got new buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("struct constructor", func(t *testing.T) {
+		t.Skip("Not yet supported.")
+		c := New()
+		buf := &bytes.Buffer{}
+		buf.WriteString("foo")
+		require.NoError(t, c.Provide(*buf), "provide failed")
+		require.NoError(t, c.Invoke(func(b bytes.Buffer) {
+			// ensure we're getting back the buffer we put in
+			require.Equal(t, "foo", buf.String(), "invoke got new buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("slice", func(t *testing.T) {
+		c := New()
+		bufs := []*bytes.Buffer{{}, {}}
```

Comment:
I'd prefer not to, since there's an ongoing discussion about whether `dig` should allow primitive values at all. I'd prefer to stick to some more-realistic type here, so that future changes don't also need to alter these high-level types.

This test should also check that invoke receives the same pointer that provide returned, though.